spec/unit/plugins/linux/network_spec.rb

```diff
@@ -1335,6 +1335,59 @@
         end
       end
 
+      describe "when there's a source field in a local route entry but no dev" do
+        let(:linux_ip_route) do
+          <<~EOM
+            default proto bird src 172.16.19.39 metric 10 \\      nexthop via 10.1.81.1 dev eth0 weight 20
+            10.116.201.0/24 dev eth0  proto kernel  src 10.116.201.76
+            192.168.0.0/24 dev eth0  proto kernel  src 192.168.0.2
+            default via 10.116.201.254 dev eth0 metric 20 src 10.116.201.74
+          EOM
+        end
+        let(:linux_ip_route_inet6) { "" }
+
+        it "maps the no-dev route and sets ipaddress" do
+          plugin.run
+          expect(plugin["ipaddress"]).to eq("172.16.19.39")
+        end
+      end
+
+      # note: default via 10.116.201.1 discarded because src not on eth0
```

Comment:
I"m not following. It has `dev eth0` in it, how do we know that 10.116.201.1 isn't the right source?  Or 172.16.19.39? How do we know that 10.116.201.74 is the "right" source?

Reply:
10.116.201.1 is the gateway IP, and not considered as source.
metric 5 is better (higher priority) than metric 10 - Hence line 1363 wins against everything else.
[The test data should've been sorted by metric, but the concept is the same]
Hence the src related to that route is chosen - 10.116.201.74

Reply:
I had missed the ifconfig at the top of the file and the code below your code that does the 'Sanity check'.

But I'm still confused. Your comment here says:

```
# note: default via 10.116.201.1 discarded because src not on eth0
```

OK, I see the code you mean:

```ruby
        # a sanity check, especially for Linux-VServer, OpenVZ and LXC:
        # don't report the route entry if the src address isn't set on the node
        # unless the interface has no addresses of this type at all
        if route_entry[:src]
          addr = iface[route_int][:addresses]
          unless addr.nil? || addr.key?(route_entry[:src]) ||
              addr.values.all? { |a| a["family"] != family[:name] }
            logger.trace("Plugin Network: Skipping route entry whose src does not match the interface IP")
            next
          end
        end
```

OK got it. But then why are any routes considered:

* `10.1.81.1` is discarded because the src `172.16.19.39` isn't on eth0
* `10.116.201.1` is discarded because the src `10.116.201.1` (same thing, nonsensicle route), isn't on eth0
* Leaving the only route left, `10.116.201.254`, but that src, `10.116.201.74` ALSO isn't on the interface, so, why is that one considered? Shouldn't that also be discarded?

The routing code here is pretty dense, each of these test cases needs a pretty clear comment on why each route is considered or not, and why of the ones that are considered, one wins (usually just "metric is lowest").